File: problems/p183.py
from time import perf_counter
from math import e
start = perf_counter()

def is_terminating(num, den):
    while den % 2 == 0:
        den //= 2
    while den % 5 == 0:
        den //= 5
    return den == 1

def D(N):
    k = round(N / e)
    if is_terminating(N, k):
        return -N
    return N

ans = 0
for n in range(5, 10001):
    ans += D(n)

print('Solution:', ans)


end = perf_counter()
print(end - start, 'seconds to run')

File: problems/p183.c
#include <stdio.h>
#include <time.h>
#include <math.h>

double E;
int D(int n) {
    int k =  round(n / E);

    while (k % 2 == 0)
        k /= 2;
    while (k % 5 == 0)
        k /= 5;

    if (k == 1)
        return -n;
    return n;
}

int main(int argc, char *argv[]) {
    clock_t start = clock();

    E = exp(1.0);
    long sum = 0;
    for (int n = 5; n <= 1000000; n++) {
        sum += D(n);
    }
    printf("Solution: %lu\n", sum);

    clock_t end = clock();
    double time = (double) (end - start) / CLOCKS_PER_SEC;
    printf("%f seconds to run\n", time);
}
